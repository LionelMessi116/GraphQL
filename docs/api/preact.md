---
title: @urql/preact
order: 2
---

# @urql/preact

Please refer to [the React API section](/api/urql) for details of the Preact API.
