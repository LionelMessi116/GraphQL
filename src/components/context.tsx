import createReactContext from 'create-react-context';
import { Client } from '../types';

const context = createReactContext<Client>({} as any); // Requires default value (which is unused)

export const ContextProvider = context.Provider;
export const ContextConsumer = context.Consumer;
