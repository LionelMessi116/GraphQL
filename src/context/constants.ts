export const CLIENT = '$$_URQL';
