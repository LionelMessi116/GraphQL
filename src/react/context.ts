import createContext from 'create-react-context';
import { Client, createClient } from '../client';

// We assume some default options here; mainly not to actually be used
// but not to error catastrophically if someone is just playing around
const defaultOptions = { url: '/graphql' };

export const Context = createContext<Client>(createClient(defaultOptions));
export const Provider = Context.Provider;
export const Consumer = Context.Consumer;
