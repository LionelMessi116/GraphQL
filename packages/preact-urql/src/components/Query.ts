import { VNode } from 'preact';
import { OperationContext } from '@urql/core';
import { useQuery, UseQueryArgs, UseQueryState } from '../hooks';

export interface QueryProps<T, V> extends UseQueryArgs<V> {
  children: (arg: QueryState<T>) => VNode<any>;
}

export interface QueryState<T> extends UseQueryState<T> {
  executeQuery: (opts?: Partial<OperationContext>) => void;
}

export function Query<T = any, V = any>(props: QueryProps<T, V>): VNode<any> {
  const queryState = useQuery<T, V>(props);
  return props.children({ ...queryState[0], executeQuery: queryState[1] });
}
