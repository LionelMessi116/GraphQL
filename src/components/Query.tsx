import React, { Component, FC, ReactNode } from 'react';
import { pipe, subscribe } from 'wonka';
import { Client } from '../client';
import { Consumer } from '../context';
import { OperationContext, RequestPolicy } from '../types';
import { CombinedError, createQuery, noop } from '../utils';

interface QueryHandlerProps {
  client: Client;
  query: string;
  variables: object;
  requestPolicy?: RequestPolicy;
  children: (arg: QueryHandlerState) => ReactNode;
}

interface QueryHandlerState {
  fetching: boolean;
  data?: any;
  error?: CombinedError;
  executeQuery: (opts?: Partial<OperationContext>) => void;
}

class QueryHandler extends Component<QueryHandlerProps, QueryHandlerState> {
  private unsubscribe = noop;

  executeQuery = () => {
    if (this.unsubscribe !== undefined) {
      this.unsubscribe();
    }

    if (this.props.query === undefined) {
      return;
    }

    this.setState({
      fetching: true,
    });

    const request = createQuery(this.props.query, this.props.variables);

    const [teardown] = pipe(
      this.props.client.executeQuery(request, {
        requestPolicy: this.props.requestPolicy,
      }),
      subscribe(({ data, error }) => {
        this.setState({
          fetching: false,
          data,
          error,
        });
      })
    );

    this.unsubscribe = teardown;
  };

  state = {
    executeQuery: this.executeQuery,
    fetching: false,
  };

  componentDidMount() {
    this.executeQuery();
  }

  componentDidUpdate(oldProps) {
    if (
      this.props.query === oldProps.query &&
      this.props.variables === oldProps.variables
    ) {
      return;
    }

    this.executeQuery();
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  render() {
    return this.props.children(this.state);
  }
}

type QueryProps = Exclude<QueryHandlerProps, 'client'>;

export const Query: FC<QueryProps> = props => (
  <Consumer>{client => <QueryHandler {...props} client={client} />}</Consumer>
);
