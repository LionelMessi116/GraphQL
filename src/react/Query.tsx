import React, { Component, FC, ReactNode } from 'react';
import { pipe, subscribe } from 'wonka';
import { Client, CombinedError, createQuery } from '../lib';
import { Consumer } from './context';

interface QueryHandlerProps {
  client: Client;
  query: string;
  variables: object;
  children: (arg: QueryHandlerState) => ReactNode;
}

interface QueryHandlerState {
  fetching: boolean;
  data?: any;
  error?: CombinedError;
}

class QueryHandler extends Component<QueryHandlerProps, QueryHandlerState> {
  private unsubscribe?: () => void;

  public state = {
    fetching: false,
  };

  public componentDidMount() {
    this.executeSubscription();
  }

  public componentDidUpdate(oldProps) {
    if (
      this.props.query === oldProps.query ||
      this.props.variables === oldProps.variables
    ) {
      return;
    }

    this.executeSubscription();
  }

  public componentWillUnmount() {
    if (this.unsubscribe !== undefined) {
      this.unsubscribe();
    }
  }

  public render() {
    return this.props.children(this.state);
  }

  private executeSubscription() {
    if (this.unsubscribe !== undefined) {
      this.unsubscribe();
    }

    if (this.props.query === undefined) {
      return;
    }

    this.setState({
      fetching: true,
    });

    const [unsubscribe] = pipe(
      this.props.client.executeSubscription(
        createQuery(this.props.query, this.props.variables)
      ),
      subscribe(({ data, error }) => {
        this.setState({
          fetching: false,
          data,
          error,
        });
      })
    );

    this.unsubscribe = unsubscribe;
  }
}

type QueryProps = Exclude<QueryHandlerProps, 'client'>;

export const Query: FC<QueryProps> = props => (
  <Consumer>{client => <QueryHandler {...props} client={client} />}</Consumer>
);
