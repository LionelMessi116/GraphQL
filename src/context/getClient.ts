import { getContext } from 'svelte';
import { Client } from 'urql/core';
import { CLIENT } from '../constants';

export const getClient = (): Client => getContext(CLIENT);
