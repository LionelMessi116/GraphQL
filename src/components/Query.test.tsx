import { h } from 'preact';
import { cleanup, render } from '@testing-library/preact';
import { map, interval, pipe } from 'wonka';
import { Query } from './Query';
import { Provider } from '../context';

const query = '{ example }';
const variables = {
  myVar: 1234,
};

const client = {
  executeQuery: jest.fn(() =>
    pipe(
      interval(200),
      map((i: number) => ({ data: i, error: i + 1 }))
    )
  ),
};

describe('Query', () => {
  beforeEach(() => {
    // eslint-disable-next-line no-console
    console.log(
      'supressing console.error output due to react-test-renderer spam (hooks related)'
    );
    jest.spyOn(global.console, 'error').mockImplementation();
  });

  afterEach(() => {
    cleanup();
  });

  it('Should execute the query', done => {
    let props = {};
    const Test = () => <p>Hi</p>;
    const App = () => {
      // @ts-ignore
      return h(Provider, {
        value: client,
        children: [
          h(
            // @ts-ignore
            Query,
            { query, variables },
            ({ data, fetching, error }) => {
              props = { data, fetching, error };
              return h(Test, {});
            }
          ),
        ],
      });
    };
    render(h(App, {}));
    expect(props).toStrictEqual({
      data: undefined,
      fetching: true,
      error: undefined,
    });
    setTimeout(() => {
      expect(props).toStrictEqual({ data: 0, fetching: false, error: 1 });
      done();
    }, 200);
  });
});
