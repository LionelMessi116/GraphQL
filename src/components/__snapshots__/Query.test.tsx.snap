// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on init default values match snapshot 1`] = `
Object {
  "data": 1234,
  "error": undefined,
  "executeQuery": [Function],
  "fetching": false,
}
`;
